apis/Google.Pubsub.V1/Google.Pubsub.V1/ResourceNames.cs

```diff
@@ -0,0 +1,194 @@
+﻿// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+
+namespace Google.Pubsub.V1
+{
+    // Resource name classes. Questions still to decide:
+    // - Should we support service-qualified resource names? If so, how?
+    // - Should we expose the PathTemplate?
+    // - Should we have an interface so that APIs which take "any resource name" can do so?
+    // - Should these be structs instead?
+
+    /// <summary>
+    /// Resource name for a project.
+    /// </summary>
+    public sealed class ProjectName
```

Comment:
Support invalid name parsing?

Have an IsValid property.
Parse method throws by default, but can be configured to not throw on invalid name - instead set IsValid to false, and throw if any ID is accessed.